src/main/java/com/google/devtools/build/lib/bazel/rules/python/python_stub_template.txt

```diff
@@ -332,54 +455,55 @@ def Main():
   if python_program is None:
     raise AssertionError('Could not find python binary: ' + PYTHON_BINARY)
 
-  cov_tool = os.environ.get('PYTHON_COVERAGE')
-  if cov_tool:
-    # Inhibit infinite recursion:
-    del os.environ['PYTHON_COVERAGE']
-    if not os.path.exists(cov_tool):
+  # COVERAGE_DIR is set iff the instrumentation is configured for the
+  # file and coverage is enabled.
+  if os.environ.get('COVERAGE_DIR'):
+    cov_tool = FindCoverageTool(module_space)
+    if not cov_tool:
+      if 'PYTHON_COVERAGE_TARGET' in os.environ:
```

Comment:
We should probably get rid of this particular hack if we're going to allow providing the coverage tool through the runtime. Having that path forwards is sufficiently nice that, assuming we get proper support from the python rules, it can just supersede the old behavior. We'll probably want to remove the old env variable in the long term.

The current implementation has some annoying edge cases, and nobody has yet suggested an alternative implementation that wouldn't. It's a footgun and introduces long-term tech debt. In fact, this hack is why my old PR was a draft.

My commits should work fine with your changes if you simply rebase on top of 72a0f9ed13e4b1f561aee4268e4919b84ac04395 instead (as an aside, avoid making merge commits on a feature branch :) ).

Reply:
Merge commits have the advantage of retaining proper attribution for the original commits.  I'd probably rebase/squash before promoting this from a draft in any case.

Yes, adding this `PYTHON_COVERAGE_TARGET` environment variable is unnecessary (and probably a bad idea to introduce)  - I was mainly just trying to reduce the diff between your changes and mine for now.  Thus, still a draft :-).  I think `PYTHON_COVERAGE` needs to stay for backwards compatibility but longer-term I think maybe the solution there would be to have a `coverage_tool_path` option on `py_runtime` similar to how there's an `interpreter_path` option for non-hermetic python toolchains.  That said, I'm not particularly interested in working on features to support non-hermetic toolchains, since I don't use them.  Apparently Google does?  But they have a lot more engineering time to spare than we do.

Reply:
> Merge commits have the advantage of retaining proper attribution for the original commits.

So do rebases, as long as you don't need to squash, which I assumed you wouldn't need to, but ignore me; obviously you would have, it touches the same code. I should do more than glance over things, but as you say later, Friday evening (well past midnight, too, in my case).

> That said, I'm not particularly interested in working on features to support non-hermetic toolchains, since I don't use them. Apparently Google does?

As far as I can tell, most, or at least a significant portion of, organizations that use Bazel still use non-hermetic toolchains - just look through public Bazel repositories. There are a lot of users that haven't yet adopted toolchains, and for which changing that would be a significant amount of effort. For the time being we'll have to play nice.

Either way, changes as-is look pretty ok then, treat this comment as a "please remove that commit before you resolve draft status", and thanks for taking this all forward :)